### PR TITLE
Adding a resource to manage Glue Records.

### PR DESCRIPTION
**Pull request by twcurrie**
Thanks to the authors for publishing this provider; its been useful for my projects to have this tool.  I have not contributed to a terraform provider, and have only contributed two previous PR's to projects written in Go (both to support contributions within this repo).  I have made an effort to follow the conventions that I see in the other resources, but do defer to the maintainers for guidance should I have not captured them correctly.

### CODE REVIEW

File: go.mod
Comment (twcurrie) [inline go.mod:7]: Ref: https://github.com/go-gandi/go-gandi/commit/d06517bcd2ac06a0f6459a57f6ac51b78075c6fe

File: docs/resources/gluerecord.md
Comment (twcurrie) [inline docs/resources/gluerecord.md]: After authoring this, I've realized that a data record would be useful, but I think I'd need to add a get request in the glue record in domain.go to request specific glue records, rather than [list all.](https://github.com/go-gandi/go-gandi/blob/master/domain/domain.go#L91-L94)

File: docs/resources/gluerecord.md
Comment (twcurrie) [inline docs/resources/gluerecord.md:1]: After authoring this, I've realized that a data record would be useful, but I think I'd need to add a get request in the glue record in domain.go to request specific glue records, rather than [list all.](https://github.com/go-gandi/go-gandi/blob/master/domain/domain.go#L91-L94)

Comment (nlewo) [reply docs/resources/gluerecord.md:1]: Yes, you need to add a Get request.

Comment (nlewo) [reply docs/resources/gluerecord.md:1]: Note the data part could also come in anther PR i guess.

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:33]:
```suggestion
				Description: "Host name of the glue record",
```

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:39]:
```suggestion
				Description: "List of IP addresses",
```

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:68]: You have to use a ["retry"](https://www.terraform.io/docs/extend/resources/retries-and-customizable-timeouts.html#retry) instead.

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:80]: You should add a `GetGlueRecords` method in the go-gandi lib instead. 

This would remove the 10lines below.

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:123]: Hm, since the `ForceNew` is set on the `values` attribute, i think Terraform will destroy and recreate the glue record on a change.

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:131]: You should also use a "retry" i guess.

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:148]: This is not required.

**Review by nlewo — CHANGES_REQUESTED** [review]
Hi @twcurrie!

Here is my review. Otherwise, this basically looks good to me ;)
Thanks for you contribution!

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go:80]: Agreed.  Added here: https://github.com/go-gandi/go-gandi/pull/47

I'll update this PR when that gets merged to simplify this code and add the "data" object for the glue_record.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go:68]: Ok, will update.  Followed the example of [this resource](https://github.com/go-gandi/terraform-provider-gandi/blob/master/gandi/resource_dnsseckey.go#L66-L67) - I suppose that should be updated?

With how this retry will work, I will also defer until I have the Get request available within the go client.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go:123]: Ahh, didn't understand what that [field](https://github.com/go-gandi/terraform-provider-gandi/pull/73/files#diff-ffa57e2a4af1be6b765bc7cbc1e1a9759533c5e6c9dc9a3330223ea4327131ffR32) [meant](https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#forcenew).

File: gandi/resource_glue_record.go
Comment (twcurrie) [inline gandi/resource_glue_record.go:146]: @nlewo This is a first stab at a retry.

Comment (nlewo) [reply gandi/resource_glue_record.go:146]: Cool, lgtm!

File: gandi/resource_glue_record.go
Comment (tiramiseb) [reply gandi/resource_glue_record.go:95]: that is merged ;)

File: docs/resources/gluerecord.md
Comment (twcurrie) [reply docs/resources/gluerecord.md:1]: Updated to integrate the Get request as opposed to the list all.  Added the data part as well for completeness.

File: gandi/data_gluerecord.go
Comment (twcurrie) [inline gandi/data_gluerecord.go:22]: Unsure if the maintainers would like to use consistent language across the dataSource ("ips" here) and resource ("values" [below](https://github.com/go-gandi/terraform-provider-gandi/pull/73/files#diff-ffa57e2a4af1be6b765bc7cbc1e1a9759533c5e6c9dc9a3330223ea4327131ffR36))

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go]: Provided.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go]: Removed.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go]: Updated.  Question around the lack of retry for dns key resource remains.

File: go.mod
Comment (twcurrie) [reply go.mod]: Updated to library integrating Get request.

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:39]: Useless since it is the default value, excepting if you really want to be explicit (up to you).

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go:36]: I was copying what the `gandi_livedomain_record` was doing, but I agree - it should be consistent with the API's terminology.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go:137]: They're identical - would you want this method defined within this file or in another file?  It appears that there's other use cases for resource retries within this library and it could be generalized as a function that accepts a callable.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go]: Not a strong preference, removed.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go:69]: I initially followed the example docs without much thought, but based on the [docs](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#Retry), I'm assuming the intention of that `time` is to just offset the timeout by a quantity.  Maybe they wanted to quicken the retry interval by a minute?  It's probably unnecessary and it's unclear why they provided in that [example](https://www.terraform.io/docs/extend/resources/retries-and-customizable-timeouts.html#retry).  

On a somewhat related note, that retry approach appears to be deprecated in favor of [`RetryContext`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#RetryContext), should that be used in favor `Retry`?

File: gandi/resource_glue_record.go
Comment (nlewo) [reply gandi/resource_glue_record.go:137]: I would be ok with a function local to this file (mainly to limit the scope of this PR).

File: gandi/resource_glue_record.go
Comment (nlewo) [reply gandi/resource_glue_record.go:69]: So, i'm ok with using `Retry`.

File: gandi/resource_glue_record.go
Comment (twcurrie) [reply gandi/resource_glue_record.go]: Added.

File: docs/resources/gluerecord.md
Comment (nlewo) [inline docs/resources/gluerecord.md:11]:
```suggestion
    ips = ["127.0.0.2"]
```

File: docs/resources/gluerecord.md
Comment (nlewo) [inline docs/resources/gluerecord.md:19]:
```suggestion
- `ips` - (Required) A list of the ip addresses of the record.
```

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:52]:
```suggestion
	for _, i := range d.Get("ips").([]interface{}) {
```

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:51]: (optional) could be renamed to `ips`

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:96]:
```suggestion
	if err = d.Set("ips", found.IPs); err != nil {
```

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:136]:
```suggestion
	if d.HasChanges("ips") {
```

File: gandi/resource_glue_record.go
Comment (nlewo) [inline gandi/resource_glue_record.go:137]:
```suggestion
		ips := d.Get("ips").([]string)
```

Comment (nlewo) [inline gandi/resource_glue_record.go:137]: Oarf, pretty sure with stuff like generics we could have detected these issues at compile time :/

**Review by nlewo — CHANGES_REQUESTED** [review]

File: gandi/resource_glue_record.go
Comment (nlewo) [reply gandi/resource_glue_record.go]: Thank you for this refactoring!

**Comment by tiramiseb** [comment]
lgtm

@nlewo ?

**Comment by nlewo** [comment]
@tiramiseb I think there are still several issues because of the switch from `values` to `ips`: i don't think the provider is currently working.

But that's minor changes, and otherwise, this lgtm.
Once fixed, i will approve.

**Comment by twcurrie** [comment]
> @tiramiseb I think there are still several issues because of the switch from `values` to `ips`: i don't think the provider is currently working.
> 
> But that's minor changes, and otherwise, this lgtm. Once fixed, i will approve.

From what I can see in the Github Action, it looks like this is just formatting of the `.md` files: 
image

I've updated them and pushed that up.  As I can not trigger the action on my side, please let me know what further changes are needed.

**Review by nlewo — APPROVED** [review]
Thank you @twcurrie !

**Comment by nlewo** [comment]
Oarf, i forget to squash your commits :(

**Comment by twcurrie** [comment]
> Oarf, i forget to squash your commits :(

May my forays into Golang forever live on in your commit history in infamy.

**Pull request merged**
